### DISCUSSION TITLE
Transcoding - the other way round?

### DISCUSSION THREAD

**Discussion by udippel** [gerbera/gerbera#2900] [General]
Having 2000+ videos, collected over 20 years, they come in many different formats. My second main monitor (SONY) plays only MPEG2, MP4, WMV. So I'm happy with transcoding, also since I don't want to convert all files. Some might lose their subtitles, what not.
Almost daily, I notice (still) missing files, add mimetypes, process mimetypes, and so forth. While this works principally, by now I'd very much prefer to work just the other way round: instead of adding more and more mimetypes for transcoding (and maybe still miss some), I'd rather send the playable ones only (.mp4, .wmv, .mpeg) directly to the drain, and then transcode all others, like
`<transcode mimetype=video/*' using="all2mpeg2"/>`

Would anyone have an idea, how to achieve that?

**Comment by KarlStraussberger** [comment]
The easiest way: update the code. Already prepared the lines. Only need to update the docs as well.

**Reply by udippel** [reply]
Only if you don't mind: What's the syntax for the 'not-to-be-coded' mime-types/containers? And what's the syntax for the 'catch-anything-else'?

Conceptually, will it be like this?:

![Transcode_generic_gerbera](https://github.com/gerbera/gerbera/assets/50265752/b3def397-f35b-4d92-a41f-66e3eec663b9)

Something like this, I'd hope. Then one could remove dozens of lines from config.xml, and rather write something to the concept of
```
<transcoding enabled="yes">
    <mimetype-profile-mappings>
      <transcode mimetype="audio/mp2" using="none" />
      <transcode mimetype="audio/mp3" using="none" />
      <transcode mimetype="audio/*" using="audio2mp3" />
      <transcode mimetype="video/mp4" using="none" />
      <transcode mimetype="video/mkv" using="none" />
      <transcode mimetype="video/*" using="video2mpeg" />
     </mimetype-profile-mappings>

```
I'm aware that XML isn't necessarily sequential. So it will need some tweaking, but will simplify the usage considerably.

**Reply by KarlStraussberger** [reply]
This looks good, but we do not know whether the particular device can play a media type. In fact that is the purpose of the `dlna-profile` element in `profile` to convince the device that the stream can be read!

In fact with the approach I added we are going th have 2 or more transcoding streams. If the device understands that at all it may pick the stream with the profile it support.

**Reply by udippel** [reply]
I understand. Once we have a version that transcodes anything that a client doesn't explicitly state to support, I'd be curious to try with the old SONY. 
The misery until now was, that anything it thinks that it can't play the format, it doesn't list the clip. Therefore I had to add a bunch of MIME-types, and a bunch of those into the mappings. And I might still lose some. Therefore I was asking for "another way round": transcoding whatever shxx it gets, except those that it actually can handle.

**Reply by KarlStraussberger** [reply]
Normally the Device picks the stream/resource it is able to play. If there is none it does not even list it. Normally 😄

**Reply by udippel** [reply]
Quite. 
Now, what's the syntax to use the newly coded feature?

**Reply by KarlStraussberger** [reply]
It's exactly the way you requested it. The consequence is the device will show the option to pick the stream (if it supports this already)

**Reply by udippel** [reply]
We seem to be running into a misunderstanding here:

```
`2023-11-28 22:35:41  error: Error in configuration: you specified a mimetype to transcoding profile mapping, but the profile "none" for mimetype "video/mp4" does not exists.

$ gerbera -v
Gerbera UPnP Server v1.12.1-81-g0966bf52`
```
If I understood you right, you also spoke about a transcoding such that any format playable by the drain would 'just play'. If this is what you coded, alas, here it doesn't work. With the `<transcode mimetype="video/*" using="video2mpeg" />` it actually DOES transcodes all and any file.

**Reply by udippel** [reply]
Why not trying for a catch-all?
The unifying "video/*" is already good progress. Suddenly all files show up on the list of video media.
In case of proper communication, that'd be all.
In case gerbera can't receive proper capabilities, or just for user-intended manual override, we offer the additional 'profile' "none" as reserved profile that doesn't require definition and doesn't do any transcoding.

How's that?

**Reply by KarlStraussberger** [reply]
That's not necessary. We send a DLNA-Profile-String so that the device can decide which stream to pick. The server does not know about capabilities of the device. So if the stream is playable by the device (show already now) it will pick that.

**Reply by udippel** [reply]
Again, BubbleUPnP does, so I think. Since it can play anything with VLC. Except I explicitly instructed otherwise. Works perfectly okay.
But that was never the problem, My main reproducing device, that old SONY, somehow plays anything transcoded, including the files it had played happily without any transcoding earlier. Half of my collection is .mp4, and the SONY officially plays MP4, and does play the MP4 presented through gerbera. 
Unfortunately, now it plays these files also as transcoded files. Seemingly, gerbera and SONY don't manage to communicate properly, so that gerbera doesn't know that the SONY does play certain files. 

Since transcoding tends to add artifacts, I'd still rather have a way to instruct gerbera to play specific files as such.

**Reply by KarlStraussberger** [reply]
There is no such thing as an exchange of capabilities.
Gerbera provides a list of resources to play and the device picks it. That's it no way of changing the devices behaviour from the servers perspective!

I know there are media servers around that maintain a list of supported file types by device. That's really bothersome business. But these servers do not support transcoding.

Supplying an additional exclusion pattern like `no-transcoding="video/mp4,video/mkv"` (list of all mime type the device supports)
In that case you have to provide multiple `<transcoding/>` items for different devices

**Reply by udippel** [reply]
You ARE right, yes. 
The missing point was me reading up on the topic of resources. The profile needs to look like this, of course:
```
<first-resource>yes</first-resource>
<hide-original-resource>no</hide-original-resource>
```
My excuses for having lacked in this respect. Done, thanks again, closed!

**Reply by KarlStraussberger** [reply]
Anyway, the exclusion list is a good idea and on its way.

**Reply by udippel** [reply]
For completeness and documentation: the flowchart further up does not reflect the current situation, rather the one with exclusions.
The current state then rather looks like this, I guess?:

![Transcode_all_gerbera](https://github.com/gerbera/gerbera/assets/50265752/64fb1fd3-75c4-4054-a3df-a966d998ee99)

And if so, any mpeg/MPEG2 would be transcoded, too? (Okay, I could try myself, but for the moment it suffices to ask for purposes of awareness.)

**Reply by KarlStraussberger** [reply]
Yes, the only additional option we have is the client flags which, if automatically detected, can be used to filter. But there is no way to directly check support.

**Reply by udippel** [reply]
Not quite!
We do know both: from 'profile' we know the target MIME type, and we know the incoming source MIME type. 
Therefore we could circumvent transcoding if both are identical!

With e.g. source being video/avi and target being video/avi and a  transcode rule of video/* we could - maybe even should - skip transcoding.
My preferred version, as always: optional.
`<transcode-to-self>{yes|no}<transcode-to-self/>`
Reason: it could be about compatibility, like in my case, and "no".
It could as well be about a specific format for the drain: bitrate, SAR, yuv, etc., and "yes".

**Comment by udippel** [comment]
Never mind this is closed, in order to add to the documentation:

Now it is 'really completed' for me. It does exactly what I had in mind. I am now running gerbera ('production', so to say) with
```
<mimetype-profile-mappings>
      <transcode mimetype="audio/*" using="audio2mp3" />
      <transcode mimetype="video/*" using="video2mpeg" /> 
</mimetype-profile-mappings>
```
And as far as I can make out, all sorts of files list/show/play on a TV with very limited reproduction capabilities; and all with a few lines.
